content/2017.04.getting-started-with-elm.rst

```diff
@@ -0,0 +1,487 @@
+
+    main =
+        Html.text "Hello world!"
+
+If not, you can run ``elm-format`` manually on your file:
+
+.. code-block:: bash
+
+    ~/tutorial$ elm-format Main.elm
+    This will overwrite the following files to use Elm's preferred style:
+
+        Main.elm
+
+    This cannot be undone! Make sure to back up these files before proceeding.
+
+    Are you sure you want to overwrite these files with formatted versions? (y/n)
+
+    y
+    Processing file Main.elm
+
+To avoid having to answer yes each time you can use the ``--yes`` parameter::
+
+    elm-format --yes Main.elm
+
+
+Opening your app in the browser
+-------------------------------
+
+One way to run your app is to use ``elm-reactor``.
+
+.. code-block:: bash
+
+    ~/tutorial$ elm-reactor
+	elm-reactor 0.18.0
+    Listening on http://localhost:8000
+
+Then open: http://localhost:8000/Main.elm
+
+You are now viewing your Elm app live!!!
+
+
+Learning about the Elm virtual DOM
+==================================
+
+The Elm Virtual DOM, functions to generate HTML are in the ``Html``
+module.
+
+If you want you can import all of it in your module using:
+
+.. code-block:: elm
+
+    import Html exposing (..)
+
+You can then use ``text`` directly for instance:
+
+.. code-block:: elm
+
+    main = text "Hello world"
+
+The Virtual DOM HTML nodes take two parameters:
+
+- A list of attributes for the node
+- A list of children for the node
+
+If I want to create a ``div`` with a link it would look like this:
+
+.. code-block:: elm
+
+    module Main exposing (..)
+
+    import Html exposing (..)
+    import Html.Attributes exposing (..)
+
+
+    main =
+        div
+            [ class "container" ]
+            [ a
+                [ href "http://www.servicedenuages.fr/" ]
+                [ text "Blog" ]
+            ]
+
+We can also create a list of links in our div:
+
+.. code-block:: elm
+
+    module Main exposing (..)
+
+    import Html exposing (..)
+    import Html.Attributes exposing (..)
+
+
+    main =
+        div
+            [ class "container" ]
+            [ ul
+                [ class "links" ]
+                [ li
+                    []
+                    [ a
+                        [ href "http://www.servicedenuages.fr/" ]
+                        [ text "Blog" ]
+                    ]
+                , li
+                    []
+                    [ a
+                        [ href "http://www.elm-lang.org/" ]
+                        [ text "Elm lang" ]
+                    ]
+                ]
+            ]
+
+
+Adding some state
+=================
+
+Now that you know how to render your page in HTML, it is time to start
+a program that knows how to handle events.
+
+The way Elm handles that is by having:
+
+- a ``Model``, an Elm record: a bit like a JavaScript object with
+  properties, that keep the state of the app,
+- an update function that will handle all the app events
+- a view function that will return the Virtual DOM that matches the state.
+
```

Comment:
At some point, it's probably worth mentioning that redux has been heavily inspired by Elm. Basically `update` is a reducer.